ability to use different ocrengine as enum, update of example.py to include ocrengine, clean up of README.md, include .vscode/ in .gitignore, update setup.py to require requests

NegassaB opened the pull request:
ability to use different ocrengine as enum, update of example.py to include ocrengine, clean up of README.md, include .vscode/ in .gitignore, update setup.py to require requests

NegassaB committed:
Add OCR engine as an Enum

.gitignore
- update to include .vscode/ files

main.py
- import Enum and unique from enum module to declare for the new
OCREngine_VAL enum class
- declare OCREngine_VAL enum class and insert two values, engine_1 which
is equal to 1 and engine_2 which is equal to 2
- use the declared enum class in API.__init__, make engine_1 default
- check if the passed value for ocrengine is an instance of
OCREngine_VAL (to enforce usage of the enum class) and the value is no
different from 1 or 2.
- use the ocrengine variable inside payload

example.py
- update to use the OCREngine_VAL enum class when wanting to change the
ocrengine of the api

README.md
- clean up the README.md file with spaces and well defined programming
languages for the code blocks
- insert a description on how to change the ocrengine with example

NegassaB committed:
correct typo

main.py
- correct the typo when checking if it's an instance of OCREngine_VAL

NegassaB committed:
Update example.py to include usage of engine_two

main.py
- give better description when throwing exception

example.py
- include example using api_with_engine_two

NegassaB committed:
Update raised error types

main.py
- change the error raised for isinstance() failure from ValueError
TypeError
- change the error raise for wrong value from Exception() to
ValueError()

NegassaB committed:
Merge branch 'add_ocrengine'

NegassaB committed:
Update setup.py to include the needed package

setup.py
- include the dependecy needed for the lib inside setup()

NegassaB committed:
update requirements.txt to list out all that are needed

NegassaB committed:
Timout exception, update README

README.md
- update README to describe getting apikey from Free OCR API.

main.py
- refactor processing of files into it's own method named query_api().
- pass timeout parameter to request.post() inside query_api() to
specifiy for how to wait before raise a Timeout error instead of hanging
the process for undefined amount of time.
- call query_api() in all the ocr processors.

NegassaB committed:
Merge branch 'optimize'

NegassaB committed:
update README.md

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
 
 
+@unique
+class OCREngine_VAL(Enum):
```
Can we call this just `OCREngine`? Or just `Engine`?

NegassaB replied:
changed to Engine on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
+    OCRengine_VAL: the OCRengine to use, values can only be 1 or 2
+
+    Args:
+        Enum: python's generic enumeration that is used to define new enumerations.
```
No need to include this as the notion of inheritance is self documenting

NegassaB replied:
removed from the docs on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
+@unique
+class OCREngine_VAL(Enum):
+    """
+    OCRengine_VAL: the OCRengine to use, values can only be 1 or 2
```
Simply replace this with `Enum representing the OCR engine to use`.

NegassaB replied:
replaced on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
+    Args:
+        Enum: python's generic enumeration that is used to define new enumerations.
+    """
+    engine_1 = 1
```
Shouldn't these be capitalized since they are constants?

NegassaB replied:
capitalized on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
         endpoint='https://api.ocr.space/parse/image',
         api_key='helloworld',
         language=Language.English,
+        ocrengine=OCREngine_VAL.engine_1,
```
Rename this parameter to `engine`

NegassaB replied:
renamed on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
         """
+        if not isinstance(ocrengine, OCREngine_VAL):
+            raise TypeError(
+                "ocrengine must be an instance of OCREngine_VAL enum class"
```
Change names here too and remove "enum class"

NegassaB replied:
changed & removed on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
+            )
+        if ocrengine.value != 1 and ocrengine.value != 2:
+            raise ValueError(
+                "the value of ocrengine must be either 1 or 2, import & use ocrspace.OCREngine_VAL"
```
Change names here as well

NegassaB replied:
changed on commit https://github.com/ErikBoesen/ocrspace/commit/77f17fb8d05996861998827e5fac2838db2b4850

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,6 +81,41 @@ def _parse(self, raw):
+                )
+            r.raise_for_status()
+        except requests.exceptions.Timeout as time_out:
+            raise time_out
```
Why bother with these `except`s? All you're doing is raising the error anyway. I think you can fully omit them unless I'm missing something you're trying to do.

NegassaB replied:
I've been using the lib for a few days now and sometimes a ```Timed out waiting for results``` occur. I believe a clearer way of handling this error is appropriate in which the user of the lib can decide what to do when any error arises.
![dfa564066f6188a1b685a067dd9e0f9aa2d250ba](https://user-images.githubusercontent.com/19590937/148231398-4e014617-e620-4f8f-b660-25847c79b72e.png)

ErikBoesen replied:
Yeah, that exception does occur, but what I'm saying is that that exception will already be raised without this `except`. You're not doing anything different with the exception at this point. If you were, it would make sense to have an `except` for it, but otherwise it's just useless code.

NegassaB replied:
done on commit https://github.com/ErikBoesen/ocrspace/commit/96a7654cb060145bd4e3c4798326cb008b8c72c6

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,6 +81,41 @@ def _parse(self, raw):
             raise Exception(raw['ErrorMessage'][0])
         return raw['ParsedResults'][0]['ParsedText']
 
+    def query_api(self, url_data=None, pic_file=None):
```
Can you think of a clearer name than `url_data`?

NegassaB replied:
how about ```image_path``` or ```image_codec```?

NegassaB replied:
anything on this? should it be ```image_path``` or ```image_codec``` or ```image_codex```? I'm lost for names.

ErikBoesen commented on requirements.txt:
```diff
@@ -1 +1,6 @@
+idna==3.3
+pycodestyle==2.8.0
+requests==2.26.0
+urllib3==1.26.7
```
It doesn't seem like we're using most of these packages, and the ones that we are are mostly dependencies of `requests` anyway.

NegassaB replied:
yes, both ```idna``` & ```urllib3``` are the dependencies of ```requests```. ```pycodestyle``` is the linter I use with vscode. All these were listed when I pip freezed.

ErikBoesen replied:
Just list `requests` then. `pycodestyle` is needed in your environment, but not necessarily to someone else trying to use the package.

NegassaB replied:
done on commit https://github.com/ErikBoesen/ocrspace/commit/96a7654cb060145bd4e3c4798326cb008b8c72c6

NegassaB committed:
Per request change

main.py
- rename OCREngine_VAL to Engine
- capitalize the values inside class Engine
- rename parameter ocrengine inside __init__() to engine
- simplify raised TypeError's message
- rename parameter pic_file to image_file in query_api()

example.py
- update example.py in accordance with main.py

NegassaB committed:
Merge branch 'review'

main.py
- rename OCREngine_VAL to Engine
- capitalize the values inside class Engine
- rename parameter ocrengine inside __init__() to engine
- simplify raised TypeError's message
- rename parameter pic_file to image_file in query_api()

example.py
- update example.py in accordance with main.py

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,48 @@ class Language:
         :param endpoint: API endpoint to contact
         :param api_key: API key string
         :param language: document language
+        :param ocrengine: ocr engine to use
```
Rename here as well

NegassaB replied:
done on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

NegassaB committed:
Per 2nd review

requirements.txt
- remove unnecessarily listed dependencies
- remove specific versioning of requests

main.py
- remove unnecessary except blocks

NegassaB committed:
Per 2nd review

requirements.txt
- remove unnecessarily listed dependencies
- remove specific versioning of requests

main.py
- remove unnecessary except blocks

NegassaB committed:
Per review #2

requirements.txt
- remove unnecessarily listed dependencies
- remove specific versioning of requests

main.py
- remove unnecessary except blocks

NegassaB committed:
Merge branch 'review'

updated in par with 2nd review

NegassaB committed:
Merge branch 'review'

updated in par with 2nd review

NegassaB committed:
Merge branch 'master' of github.com:NegassaB/ocrspace

NegassaB committed:
Merge branch 'master' of github.com:NegassaB/ocrspace

ErikBoesen commented on README.md:
````diff
@@ -3,34 +3,63 @@
+```python
+import ocrspace
+
+api_key = "apikey retrieved from Free OCR API"
````
Use single quotes here for consistency

NegassaB replied:
done on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

ErikBoesen commented on README.md:
```diff
@@ -3,34 +3,63 @@
 
 ## Use
+
+Before anything else, you need to get your apikey from [Free OCR API](https://ocr.space/OCRAPI) otherwise the default apikey -- i.e **helloworld** -- will be used and according to the developers of [Free OCR API](https://ocr.space/faq#span12):
```
Replace with this content:

First, get an API key from [Free OCR API](https://ocr.space/OCRAPI), otherwise the default apikey `helloworld` will be used, which is [severely rate limited](https://ocr.space/faq#span12).

NegassaB replied:
replaced on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

ErikBoesen commented on README.md:
```diff
@@ -3,34 +3,63 @@
+api = ocrspace.API(api_key=api_key, ocrengine=OCREngine_VAL.val_2)
 # Or if you have a custom API host, API key or desired language, pass those:
-api = ocrspace.API(endpoint='https://example.host', api_key='Insert key here', language=ocrspace.Language.Croatian)
+api = ocrspace.API(endpoint='https://example.host', api_key=api_key, language=ocrspace.Language.Croatian, ocrengine=OCREngine_VAL.val_2)
```
Change names here

NegassaB replied:
changed on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,45 @@ class Language:
+@unique
+class Engine(Enum):
+    """
+    Engine: Enum representing the OCR engine to use
```
Remove `Engine: `

NegassaB replied:
removed on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,45 @@ class Language:
+            )
+        if engine.value != 1 and engine.value != 2:
+            raise ValueError(
+                "the value of engine must be either 1 or 2, import & use ocrspace.Engine"
```
Remove `import &`

NegassaB replied:
removed on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,21 +78,39 @@ def _parse(self, raw):
 
-
-    def ocr_file(self, fp):
+    def query_api(self, url_data=None, image_file=None):
```
Actually I think you can just leave `url_data`. I shy away from using `data` in names, but here I think it's appropriate.

ErikBoesen replied:
Actually, hold on—is there anything preventing us from separating this into two parameters, one for base64 image data and one for a URL?

NegassaB replied:
IMHO, it wouldn't make sense to separate them cause the query they conduct is handled by the same ```elif``` statement inside ```query_api()```
```python
elif url_data:
    r = requests.post(
        self.endpoint,
        data=url_data,
        timeout=30
    )
```

ErikBoesen replied:
Sure, but we could very easily recombine them; this would just provide syntactic sugar.

NegassaB replied:
yea but separating something just to recombine it later on...seems kinda clunky as well as inefficient

ErikBoesen replied:
I it's well worth it to have the meaning of the parameter be clear. `url_data` is pretty ambiguous and could mean a lot of things. One `or` statement to allow the meaning of both these parameters to be self-documenting is a worthy compromise.

NegassaB committed:
Add a headers parameter, move apikey to headers

main.py
- remove key api_key from payload dict
- insert new self attribute named api_key
- add a headers parameter in requests.post() and place api_key in
headers

This was done in accordance with the documentation and examples provided by Free OCR API https://ocr.space/ocrapi#postman

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,21 +78,39 @@ def _parse(self, raw):
+                self.endpoint,
+                data=url_data,
+                timeout=30
             )
```
Add an `else` to handle the case where neither `image_file` nor `url_data` are provided.

NegassaB replied:
added on commit https://github.com/ErikBoesen/ocrspace/commit/63929964b74343a113172bcbc3f3aeec5c72de65

NegassaB committed:
Merge branch 'review'

ErikBoesen commented on README.md:
````diff
@@ -3,34 +3,62 @@
 
 ## Use
+
+First, get an API key from [Free OCR API](https://ocr.space/OCRAPI), otherwise the default apikey ```helloworld``` will be used, which is [severely rate limited.](https://ocr.space/faq#span12)
````
Just use single tick marks around helloworld so that a separate line isn't created.

NegassaB replied:
done on commit https://github.com/ErikBoesen/ocrspace/commit/dbcf91afd16977d60018f2985188080ed8b987ee

ErikBoesen commented on README.md:
````diff
@@ -3,34 +3,62 @@
+api = ocrspace.API(endpoint='https://example.host', api_key=api_key, language=ocrspace.Language.Croatian)
+```
+
+If you wish to change the OCR engine used, you'll have to import the enum class Engine from ocrspace and pass the value of Engine.ENGINE_2 to the api instantiation. By default it uses Engine.val_1:
````
Replace `val_1` with new name

NegassaB replied:
done on commit https://github.com/ErikBoesen/ocrspace/commit/dbcf91afd16977d60018f2985188080ed8b987ee

NegassaB committed:
Per 3rd review

main.py
- remove Engine: from docstring for Engine
- rename ocreengine to engine in docstring for API
- remove import & from ValueError message
- update query_api() to if/elif/else statement when querying OCR api

README.md
- use clear & concise wording in the Use section
- rename OCREngine_VAL to Engine
- rename OCRENGINE_VAL.val_1 & OCRENGINE_VAL.val_2 to Engine.ENGINE_2 &
ENGINE.ENGINE_1

NegassaB committed:
Merge branch 'use_headers' into copied

NegassaB committed:
Merge branch 'use_headers'

* It might solve the TimeoutError

main.py
- remove key api_key from payload dict
- insert new self attribute named api_key
- add a headers parameter in requests.post() and place api_key in
    headers

N.B. This was done in accordance with the documentation and examples provided by Free OCR API https://ocr.space/ocrapi#postman

ErikBoesen commented on README.md:
```diff
@@ -3,34 +3,62 @@
+api = ocrspace.API(api_key=api_key, engine=Engine.ENGINE_2)
 # Or if you have a custom API host, API key or desired language, pass those:
-api = ocrspace.API(endpoint='https://example.host', api_key='Insert key here', language=ocrspace.Language.Croatian)
+api = ocrspace.API(endpoint='https://example.host', api_key=api_key, language=ocrspace.Language.Croatian, engine=Engine.ENGINE_2)
```
Since we're using `ocrspace.Language`, it makes sense to me to use `ocrspace.Engine` as well rather than importing directly.

NegassaB replied:
updated on commit https://github.com/ErikBoesen/ocrspace/commit/dbcf91afd16977d60018f2985188080ed8b987ee

ErikBoesen commented on example.py:
```diff
@@ -1,22 +1,34 @@
 import ocrspace
+from ocrspace import Engine
```
Same here

NegassaB replied:
updated on commit https://github.com/ErikBoesen/ocrspace/commit/dbcf91afd16977d60018f2985188080ed8b987ee

ErikBoesen reviewed (changes requested):
Almost ready to go!

NegassaB committed:
Per the 4th review

README.md
- use single quotes instead of triple codes for describing the apikey in the Use section
- remove description and import of Engine, use instead full package name
- update example code to use full package name for Engine

NegassaB committed:
add maintainer's name

ErikBoesen commented on LICENSE:
```diff
@@ -1,8 +1,11 @@
+# MIT License
 
 Copyright (c) 2017 Ali Najafi
+
```
Regular license format does not require these intermediary newlines

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on README.md:
```diff
@@ -3,34 +3,61 @@
 ## Authorship
-This package was created by [Ali Najafi](https://github.com/a4fr) and is maintained by [Erik Boesen](https://github.com/ErikBoesen).
+
+This package was created by [Ali Najafi](<https://github.com/a4fr>) and is maintained by [Erik Boesen] (<https://github.com/ErikBoesen>) and Negassa Berhanu (<https://github.com/NegassaB>).
```
Please un-break these links

NegassaB replied:
can you please explain? when you say "un-break these links" do you mean remove the angle brackets?

NegassaB replied:
never mind, figured out what you meant :+1:

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,45 @@ class Language:
         """
+        if not isinstance(engine, Engine):
+            raise TypeError(
+                "engine must be an instance of Engine"
```
Use single quotes for consistency also no need to make this span three lines

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,45 @@ class Language:
     Turkish = 'tur'
 
 
+@unique
```
I don't think you need to use the unique decorator here

NegassaB replied:
resolved on c9ffd78

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,45 @@ class Language:
 
 
+@unique
+class Engine(Enum):
```
Use `IntEnum` instead; that way you can directly compare with an integer and it will still work automatically

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,21 +78,44 @@ def _parse(self, raw):
+                timeout=30
+            )
+        else:
+            raise TypeError("either image_file or url_data must be provided")
```
Single quotes

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on setup.py:
```diff
@@ -1,13 +1,14 @@
+    version='2.3.0',
+    description='Perform OCR through ocr.space API',
+    author=['Ali Najafi', 'Erik Boesen', 'Negassa Berhanu'],
+    author_email='forpurposes1435@gmail.com',
```
Please leave my email in there

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c...with my apologies

ErikBoesen reviewed (changes requested):
Thanks so much for your help! Just a few more things to fix.

ErikBoesen commented on LICENSE:
```diff
@@ -1,8 +1,11 @@
-MIT License
+# MIT License
```
This isn't a markdown file

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on README.md:
```diff
@@ -3,34 +3,61 @@
+
+api_key = "apikey retrieved from Free OCR API"
+
+api = ocrspace.API(api_key=api_key, engine=ocrspace.Engine.Engine.ENGINE_2)
```
I think you duplicated `Engine.` here

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on README.md:
```diff
@@ -3,34 +3,61 @@
+api = ocrspace.API(api_key=api_key, engine=ocrspace.Engine.Engine.ENGINE_2)
 # Or if you have a custom API host, API key or desired language, pass those:
-api = ocrspace.API(endpoint='https://example.host', api_key='Insert key here', language=ocrspace.Language.Croatian)
+api = ocrspace.API(endpoint='https://example.host', api_key=api_key, language=ocrspace.Language.Croatian, engine=ocrspace.Engine.Engine.ENGINE_2)
```
And here

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,21 +78,44 @@ def _parse(self, raw):
+        :param fp: A path or pointer to your file
+        :return: Result in JSON format
+        """
+        with (open(fp, 'rb') if type(fp) == str else fp) as f:
```
This is kind of dense, can we reorganize to be a bit clearer?

NegassaB replied:
can you clarify on this. Do you want to reorganize the code or the documentation? and if so how do you want to reorganize the code or documentation? Dense doesn't really communicate what you want to do

ErikBoesen replied:
The code. As in, split the ternary if/else into a full if/else statement across multiple lines so that we don't have a ton of logic in a single line making it difficult to read and understand at a glance

jurajlutter replied:
> can you clarify on this. Do you want to reorganize the code or the documentation? and if so how do you want to reorganize the code or documentation? Dense doesn't really communicate what you want to do

The code could be made more readable (or more verbose).

ErikBoesen commented on setup.py:
```diff
@@ -1,13 +1,14 @@
+    name='ocrspace',
+    packages=['ocrspace'],  # this must be the same as the name above
+    requires=['requests'],
+    version='2.3.0',
```
Can you bump the version to 2.4.0?

NegassaB replied:
done on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -1,4 +1,5 @@
 import requests
+from enum import (Enum, unique)
```
Remove the unique import and the parentheses

NegassaB replied:
resolved on c9ffd78ddc93ede49af145fe706ee269c4e0a42c

NegassaB committed:
Updated in accordance with the 6th review

LICENSE
- remove h1 tag from header

README.md
- remove duplicates of Engine
- remove spaces from the authors' names & urls

setup.py
- bump version to 2.4.0
- restore author's email

main.py
- remove unique and paranthesis from enum import
- replace Enum import by IntEnum
- replace double quotes with single quotes
- rename url_data to image_url with a better description in the docs

NegassaB committed:
Merge branch 'release'

ErikBoesen commented on LICENSE:
```diff
@@ -1,7 +1,8 @@
-MIT License
+ MIT License
```
No need for leading space either

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -28,25 +29,40 @@ class Language:
         """
+        if not isinstance(engine, Engine):
+            raise TypeError('engine must be an instance of Engine')
+        if engine.value != 1 and engine.value != 2:
```
I don't think this check is needed; it's impossible to make an IntEnum class with an invalid value as far as I know

ErikBoesen commented on ocrspace/main.py:
```diff
@@ -57,21 +73,44 @@ def _parse(self, raw):
-        Process image from a local path.
-        :param fp: A path or pointer to your file
+        Process the provided parameter.
+        :param image_url: An Image url or base64image encoded string
```
You may have forgotten to use this parameter in the function

ErikBoesen reviewed (changes requested):
oops, wrote these comments up a long time ago and then never submitted the review. Looks like I was only looking for some small changes so should be straightforward
